### DISCUSSION TITLE
how do print a parsed schema

### DISCUSSION THREAD

**Discussion by softprops** [graphql-java/graphql-java#2666] [Q&A]
GraphQL has two very useful types `SchemaParser` which allows you to merge a set of schema types, inputs, directives, and inputs  and `SchemaPrinter` which allows you to print a schema but it's clear how to use one in term of the other. A SchemaParser returns `TypeDefinitionRegistry` whose types do not seem compatible with any of the inputs defined for `SchemaPrinter`. 

I'd like to use the combination of the two to print a merged unified schema which for me is a composition of separate schema files. 

Is this possible with todays graphql-java?

**Comment by softprops** [comment]
after a bit more digging I discovered what I think I was looking for

```java
new SchemaPrinter().print(SchemaGenerator.createdMockedSchema(rawSchema)))
```

The trick is that you need to lift a set of definitions into an executable schema and the easy way to do that without loading up all your type writings is to use the mock wirings provided.

https://github.com/graphql-java/graphql-java/blob/1ee9630c7d25ccb55a5b1d9b1ccf7bf7895a8c7f/src/main/java/graphql/schema/idl/SchemaGenerator.java#L41

**Reply by bbakerman** [reply]
Yeah so SchemaParser only generates the elements of the  "Abstract Syntax Tree" (AST).  This is not a schema.  Its the pure AST elements - it could in theory be an invalid schema

For example

```
type Query  {
    f1 : String
    f1 : Int
}
```

would parse but it would NOT be valid.

`graphql.language.AstPrinter` can print AST elements.

You then build the runtime schema from that AST elements (SchemaGenerator).  This will make sure its truly valid and resolve type references etc... and produce a `graphql.schema.GraphQLSchema` - a valid complete runtime schema.

`graphql.schema.idl.SchemaPrinter` can then print a runtime schema

**Reply by softprops** [reply]
This is good to know. Thanks for such a wonderful library!
